src/Npgsql/Internal/NpgsqlReadBuffer.cs

```diff
@@ -194,7 +188,7 @@ internal Task Ensure(int count, bool async, bool readingNotifications)
                     // In this case, we consider it as timed out and fail with OperationCancelledException on next ReadAsync
                     // Or we consider it not timed out if we have already read everything (count == 0)
                     // In which case we reinitialize it on the next call to EnsureLong()
-                    if (async)
+                    if (async && count > 0)
```

Comment:
Optimizing for `ReadBuffer` is going to be tricky. Unlike `WriteBuffer`, `ReadBuffer.Ensure` can be cancelled at any time from a different thread.

Reply:
This makes sense to me, but to go further: when count == 0 (no more to read), shouldn't we stop the timer completely? Otherwise, it will continue "ticking" and eventually fire, forcing us to create a new CancellationTokenSource etc.?

Reply:
Already do this at the bottom.

Reply:
Right, missed that.

This method - and ResettableCancellationTokenSource - are pretty complex/tricky..